### PR TITLE
Security

### PR DESCRIPTION
**Pull request by patrickSpaceSurfer**

### CODE REVIEW

File: core/src/main/java/io/questdb/cairo/map/CompactMap.java [hunk 1]
```diff
@@ -670,27 +675,24 @@ private void putEntryAt(long entryOffset, long slot, byte flag) {
...
-                    // reached capacity?
-                    // no need to populate slot, grow() will do the job for us
-                    grow();
-                } else {
-                    setOffsetAt(slot, currentEntryOffset);
-                }
-                // this would be offset of entry values
-                value.of(currentEntryOffset, true);
-                return value;
+            resourceLimiter.checkLimits(size);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/map/CompactMap.java:678]: both maps are fixed size structure most of the time, checking limits should be done on resize. We can perhaps skip initial sizing but prohibit further resize if limits do no allow.

Also limit check call will be inside hopefully non-inlined call to grow() and have 0 performance impact on most things

File: core/src/main/java/io/questdb/cairo/map/FastMap.java [hunk 2]
```diff
@@ -249,17 +250,14 @@ public MapKey withKey() {
...
-            offsets.set(index, keyWriter.startAddress - kStart);
-            if (--free == 0) {
-                rehash();
-            }
-            size++;
-            return value.of(keyWriter.startAddress, true);
-        } else {
-            throw LimitOverflowException.instance(maxSize);
-        }
+        resourceLimiter.checkLimits(size);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/map/FastMap.java:253]: same here, on resize()

File: core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java [hunk 3]
```diff
@@ -107,8 +107,7 @@ public void execute0(JsonQueryProcessorState state) throws PeerDisconnectedExcep
         final HttpConnectionContext context = state.getHttpConnectionContext();
         // do not set random for new request to avoid copying random from previous request into next one
         // the only time we need to copy random from state is when we resume request execution
-        sqlExecutionContext.with(context.getCairoSecurityContext(), null, null);
-        sqlExecutionContext.setRequestFd(context.getFd());
+        sqlExecutionContext.with(context.getCairoSecurityContext(), null, null, context.getFd(), context.getSqlExecutionInterruptor());
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java:110]: our users use questdb as a library and we had mild complaints that SqlExecutionContext keep changing from version to version. Perhaps overloading methods here can go a long way to keep users happy.

Comment (bluestreak01) [reply core/src/main/java/io/questdb/cutlass/http/processors/JsonQueryProcessor.java:110]: I'm guilty of doing this as well. I didn't mean to put you on the spot here, but we perhaps should start changing the course of the execution context :)

File: core/src/main/java/io/questdb/griffin/SqlResourceLimiterImpl.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import io.questdb.griffin.engine.LimitOverflowException;
+
+public class SqlResourceLimiterImpl implements SqlResourceLimiter {
+    private long maxInMemoryRows;
+    private SqlExecutionInterruptor interruptor;
+
+    @Override
+    public void checkLimits(long nRows) {
+        if (nRows < maxInMemoryRows) {
+            interruptor.checkInterrupted();
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlResourceLimiterImpl.java:12]: I see the reason for checks on every map operation. I guess idea doing this on map was to do it in one place? I have a disappointment here, we have other operations needing interruptions that do not involve map at all. I would suggest to decouple limit check and interrupt.

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java [hunk 5]
```diff
@@ -72,13 +73,11 @@ public RecordCursor getCursor(SqlExecutionContext executionContext) {
...
-            long maxInMemoryRows = executionContext.getCairoSecurityContext().getMaxInMemoryRows();
-            if (maxInMemoryRows > baseCursor.size()) {
-                dataMap.setMaxSize(maxInMemoryRows);
-                cursor.of(baseCursor, dataMap, mapSink);
-                return cursor;
-            }
-            throw LimitOverflowException.instance(maxInMemoryRows);
+            SqlResourceLimiter resourceLimiter = executionContext.getResourceLimiter();
+            resourceLimiter.checkLimits(baseCursor.size());
+            dataMap.setResourceLimiter(resourceLimiter);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java:78]: Execution context is bound to thread, so it resource limiter by extension. The RecordCursorFactory and its map is not thread bound. There is strong possibility that Thread A will be using resource limited of Thread B. I'm not sure if there is an issue right now doing this, but this potentially opens an unwanted can of worms in the future

File: core/src/test/java/io/questdb/griffin/TableBackupTest.java [hunk 6]
```diff
@@ -124,7 +124,7 @@ public TimestampFormat getBackupDirTimestampFormat() {
         MessageBus mainMessageBus = new MessageBusImpl();
         mainEngine = new CairoEngine(mainConfiguration, mainMessageBus);
         mainCompiler = new SqlCompiler(mainEngine);
-        mainSqlExecutionContext = new SqlExecutionContextImpl(mainMessageBus, 1, mainEngine).with(AllowAllCairoSecurityContext.INSTANCE, new BindVariableService(), null);
+        mainSqlExecutionContext = new SqlExecutionContextImpl(mainMessageBus, 1, mainEngine).with(AllowAllCairoSecurityContext.INSTANCE, new BindVariableService(), null, -1, null);
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/griffin/TableBackupTest.java:127]: Case in point. Changes to SqlExecutionContext should have very limited impact on tests that are not using new features.

File: core/src/main/java/io/questdb/griffin/HttpSqlExecutionInterruptor.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        super();
+        this.nf = nf;
+        this.nIterrationsPerCheck = nIterrationsPerCheck;
+        this.bufferSize = bufferSize;
+        buffer = Unsafe.malloc(bufferSize);
+    }
+
+    @Override
+    public void checkInterrupted() {
+        if (nIterrationsSinceCheck == nIterrationsPerCheck) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/HttpSqlExecutionInterruptor.java:27]: this method is or can be on a "hot" data path. We should pay special attention to branch prediction here and inlining. Perhaps microbench of this method's impact would be good to have.

Without microbech results I would suggest this method styles as follows:

```java
if (++nIterrationsSinceCheck < nIterrationsPerCheck) {
    return;
}
doHeavyCheck();

File: core/src/main/java/io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java (hunk 5 above)
Comment (patrickSpaceSurfer) [reply core/src/main/java/io/questdb/griffin/engine/groupby/DistinctRecordCursorFactory.java:78]: Surely this problem applies to the map as well

**Pull request closed without merging**